(@fluent/react) Add useTranslate

macabeus opened the pull request:
Closes this issue: https://github.com/projectfluent/fluent.js/issues/467

Add `useTranslate` custom hook, that uses the `FluentContext` to return two functions to get the translated messages from Fluent's bundle.
- `t`: get the root message from the given id
- `tAttributes`: get the attributes message from the given id

This PR also updates one example file on `@fluent/react` in order to show how to use that.

macabeus commented on fluent-react/example/src/Hello.tsx:
```diff
@@ -1,29 +1,23 @@
-            </Localized>
+            <input
+                type="text"
+                placeholder={tAttributes('type-name').placeholder}
```
I don't know if add a function `tAttributes` is the best way, but I think that it's enough.
At the first moment I tried adding a parameter on `t` function to return the attributes (`t('type-name'), undefined, { attributes: true })`), but it's very weird and bad to type.

macabeus commented on fluent-react/src/useTranslate.ts:
```diff
@@ -0,0 +1,96 @@
+export const useTranslate: useTranslate = () => {
+  const l10n = useContext(FluentContext);
+
+  if (!l10n) {
+    // Return fallbacks functions
+    const t: TFunction = () => "";
+    const tAttributes: TAttributesFunction = () => ({});
+
+    return { t, tAttributes };
+  }
```
I'm returning an empty string and empty object because I'm following the same approach of `<Localized />`, but maybe would be better to return the `id`.
On the case of `tAttributes`, we could use Proxy to return the attribute name. For example:

```js
tAttributes('invalid-key').foo // should return `"foo"`
```

stasm commented:
Thank you for the PR, @macabeus! I've been thinking about adding a hook-based API too, and it's great to see a contribution in this area. I'd like to spend a bit of time discussing the API design and the use-cases, and then I'll be happy to do a code review. CC'ing @Gregoor who re-wrote `@fluent/react` to function components which use hooks internally.

----

The API I had in mind was a bit different, and I wonder you think about it. Rather than return `t` and `tAttributes`, could we directly return the `ReactLocalization` instance? It already has the `getString` method which is similar to your `t` function. We could then do:

```js
let l10n = useLocalization();
alert(l10n.getString("hello"));
```

Or even:

```js
let l10n = useContext(FluentContext);
alert(l10n.getString("hello"));
```

What I like in this approach is that it mostly relies on abstractions already built into React and into `@fluent/react`. I usually try to look for solutions which avoid adding more code :)

----

I also have a suggestion regarding the `tAttributes` function. In Fluent, querying individual attributes alone shouldn't be needed. In other words, the most common operation is either querying the value of the message (like in `getString), or querying the value _and_ the attributes.

In `@fluent/dom` we have two functions corresponding to these two use-cases: `valueFromBundle` (which is functionally equivalent to `@fluent/react`'s `getString`, and `messageFromValue`, which is implemented like this:

https://github.com/projectfluent/fluent.js/blob/cff95f6f42338974e90e58607c9aaa87c92021d3/fluent-dom/src/localization.js#L215-L235

Perhaps we could add a similar method to `ReactLocalization` to allow formatting the value _and_ the attributes in one go?

----

 That said, I might suggest startig with just `getString` for now. For more complex usage, I think wrapping components in `<Localized>` should be the preferred way of localizing them together with attributes.

macabeus commented:
Thank you for commenting on the PR. And yeah, I also want to discuss more about the API =]

<hr />

> Rather than return `t` and `tAttributes`, could we directly return the `ReactLocalization` instance? It already has the `getString` method which is similar to your `t` function.

I didn't notice about `ReactLocalization`.
I'm making some prototypes following this approach, and yeah, I think that returning a `ReactLocalization` would be enough.

I like the mindset of creating a new feature with few new line of codes, but:
- I think that the name `getString` is too big. Is very common to use many times the function that translate a string, and because of that is a common convention to call this function just as `t`. A smaller name is good on this case. Maybe we could rename `getString` to `t`?
- I'm a little afraid to `useTranslate` just returning `ReactLocalization` instance instead of a plain object, because it'll would bad if we want to change its return on future to returning something more. Do you think that it could be a problem?

<hr />

> Perhaps we could add a similar method to `ReactLocalization` to allow formatting the value _and_ the attributes in one go?

Yeah, adding a new method that returns everything would be nice. Looks like a better approach following the fluent mindset.

stasm commented:
> I think that the name getString is too big. Is very common to use many times the function that translate a string, and because of that is a common convention to call this function just as t. A smaller name is good on this case. Maybe we could rename getString to t?

I think `t` is a good choice when the function is expected to be used a lot, like in i18next. For `@fluent/react`, I think we should stick to composing components via `<Localized>` and only use `getString` in exceptional scenarios, when the imperative API is the only way to solve a problem. The way I think about the API I'd like to add here is that it's a hook equivalent of the [`withLocalization`](https://github.com/projectfluent/fluent.js/wiki/withLocalization) HOC, catering to the same use-caes.

That's why I'd suggest to only focus on the `getString` use-case. For attributes, let's leave that out of this PR, please. We might want to add a method for formatting an entire message, together with its attributes, later on, but for now, I'd like to focus on `<Localized>` as the work horse.

> I'm a little afraid to useTranslate just returning ReactLocalization instance instead of a plain object, because it'll would bad if we want to change its return on future to returning something more. Do you think that it could be a problem?

Hmm, that's an interesting point. I think you're right that it might be safer to return a plain object, like so:

```js
let {l10n} = useLocalization();
```

In the future, we might want to add more properties, for example:

```js
let {l10n, changeLocales} =  useLocalization();
```

I'd like to think about this a little bit and see what's possible. Let's not block this PR on this; both approaches will be fine :)

macabeus commented:
@stasm Okay. I just changed the PR to `useTranslate` return an instance of `ReactLocalization` (and a fallback object if there isn't that, following the [null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern)).

Also I changed the example to just replace `withLocalization` with the new hook.

stasm commented on fluent-react/src/useTranslate.ts:
```diff
@@ -0,0 +1,28 @@
+import { ReactLocalization } from "./localization";
+
+// Implementation of null object pattern for ReactLocalization
+const fallbackL10n: ReactLocalization = {
```
Is this actually needed? I understand and agree with the intent, but it also seems to me that `useContext` is guaranteed to return an empty `ReactLocalization` thanks to how it's initialized here:

https://github.com/projectfluent/fluent.js/blob/c27994d75a8a708540a3b631617a271ec48253b9/fluent-react/src/context.ts#L4

macabeus replied:
Okay, I removed that. I didn't notice that we are setting a default value on `createContext`

stasm commented on fluent-react/src/useTranslate.ts:
```diff
@@ -0,0 +1,28 @@
+* The `useTranslate` hook returns the FluentContext
+*/
+type useTranslate = () => { l10n: ReactLocalization }
+export const useTranslate: useTranslate = () => {
```
Would you mind renaming this to `useLocalization`, to keep it in sync with the `withLocalization` HOC?

macabeus replied:
Okay. Makes sense to me.

stasm commented on fluent-react/src/useTranslate.ts:
```diff
@@ -0,0 +1,28 @@
+  return { l10n };
+};
+
+export default useTranslate;
```
Let's not mix default exports with named exports. In fact, I try to avoid default exports in TypeScript altogether.

macabeus commented:
@stasm I made the changes that you proposed.
Also I added some very simple tests case for `useLocalization`. I added just three new tests because I'm avoiding to add redundant tests.

stasm commented on fluent-react/test/use_localization.test.js:
```diff
@@ -0,0 +1,57 @@
+    `);
+  });
+
+  test("useLocalization should exposes getString from ReactLocalization", () => {
```
```suggestion
  test("useLocalization exposes getString from ReactLocalization", () => {
```

stasm commented:
@macabeus Let me know when you're ready for getting this merged. I'll also update the docs on the wiki once this is released. Thanks again!

macabeus commented:
@stasm It's fine to me to merge =]
Thank you
